# Heroes of Pymoli Data Analysis

See the very bottom of this page for three observable trends based on the data.

In [1]:
# Dependencies and Setup.

import pandas as pd
import numpy as np

# File to Load - purchase data.

data = "Resources/Heroes_of_Pymoli_Purchase_Data.csv"

# Read Purchasing File and store into Pandas data frame.

purchase_data = pd.read_csv(data, low_memory=False)

# Display the first 5 rows.

purchase_data.head()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price  
0  Extraction, Quickblade Of Trembling Hands   3.53  
1                          Frenzied Scimitar   1.56  
2                               Final Critic   4.88  
3                                Blindscythe   3.27  
4                                       Fury   1.44

## Player Count:


The table below will display the total number of players in the game.

In [2]:
# To get the total number of players, I need to count the number of unique user names (or "SN") in the "SN" column.
# The unique function needs to be used because some players purchased multiple items and thus show up more than once in the list.
# Counting all the "SN" listed would not show the accurate number of players because there would be duplicates.

sum_players = len(purchase_data["SN"].unique())

# To get the data I collected above to show as a data frame or table, I need to create a dictionary.
# I want the column title to be "Total Players" and the result to be the total amount of players (sum_players).
# I used pd.DataFrame() to create a data frame with the above data.

player_count = [{"Total Players": sum_players}]
player_count_df = pd.DataFrame(player_count)

#Display the formatted table with the values.

player_count_df

Total Players
0            576

## Purchasing Analysis (Total): 


The following table will display the number of unique items, the average purchase price, the total number of purchases, and the total revenue. The table has been formatted to display a cleaner table of data.

In [3]:
# To find the number of unique items, I used len() and unique() to count the number of unique values in the "Item ID" column.

unique_items_total = len(purchase_data["Item ID"].unique())

# To find the total revenue, I used sum() to find the total of all the values in the "Price" column.

total_revenue = purchase_data["Price"].sum()

# To find the total amount of purchases, I used count() to get the total number of items listed under the "Price" column.

total_purchases = purchase_data["Price"].count()

# To find the average price of items purchased in the game, I used mean() to get the average of the values in the "Price" column.

average_price = purchase_data["Price"].mean()

# To display the data collected above into a table, I created a dictionary and then converted it to a data frame.
# To get a cleaner display, I also formatted the "Average Price" and "Total Revenue" to display in $0.00 formatting.

purchasing_analysis = [{"Number of Unique Items": unique_items_total, "Average Price": "${:,.2f}".format(average_price), "Number of Purchases": total_purchases, "Total Revenue": "${:,.2f}".format(total_revenue)}]
purchasing_analysis_df = pd.DataFrame(purchasing_analysis)

# I used reindex() to re-order the columns.

purchasing_analysis_df = purchasing_analysis_df.reindex(columns=["Number of Unique Items", "Average Price", "Number of Purchases", "Total Revenue"])

#Display the formatted table with the values.

purchasing_analysis_df

Number of Unique Items Average Price  Number of Purchases Total Revenue
0                     183         $3.05                  780     $2,379.77

## Gender Demographics: 


The following table will display the percentage and count of male players, female players, and other/non-disclosed players.

In [4]:
# Using drop_duplicates(), I created a data frame with unique players only since I am trying to see the gender data.
# If I did not get rid of the duplicate players, there would be an incorrect total as some players bought multiple items.

unique_sn = purchase_data.drop_duplicates(["SN"])

# Using value_counts(), I found the total of each gender based on the data frame created above.

gender_counts = unique_sn["Gender"].value_counts()

# I want my data displayed as a table so I converted the data above into a data frame.

gender_count_df = pd.DataFrame(gender_counts)

# Since the data frame will now display the column name as "Gender," I need to rename the column to "Total Count."

gender_count_df = gender_count_df.rename(columns = {"Gender" : "Total Count"})

# I need a new column called "Percentage of Players" that has the percentage of each gender type listed accordingly.
# Using a list comprehension, I found the percentage of each gender by dividing the "Total Count" value by the total amount of players found earlier(sum_players).
# I also added formatting to the values to do everything at once.

gender_count_df["Percentage of Players"] = ["{:.2f}".format((i/sum_players)*100) for i in gender_count_df["Total Count"]]

# Display the formatted table with the values.

gender_count_df

Total Count Percentage of Players
Male                           484                 84.03
Female                          81                 14.06
Other / Non-Disclosed           11                  1.91

## Purchasing Analysis (Gender): 


The following table will display the purchase count, average purchase price, total purchase value, and average purchase total per person for each gender.

In [5]:
# To get my final table, I need to create a data frame to start with first.
# Since the first column should be "Purchase Count", I used groupby() to group "Item Name" by "Gender." I can actually choose any column to count since I am just getting a total. I chose "Item Name.
# I did not use value_counts() here because I want "Female" to be listed as the first index. The reason for this is because when I use groupby for the price averages and totals later, I want to keep the indexes the same so I can easily combine the data.
# So the data frame I am creating will be called "purchase_analysis_by_gender_df" and I am creating the data frame by using pd.DataFrame on the data I collected by using the groupby() of the "Item Name" by "Gender."
# Since the column name will be "Item Name," I will need to use rename() to change the name from "Item name" to "Purchase Count"

purchase_analysis_by_gender_df = (pd.DataFrame(purchase_data.groupby(["Gender"]).count()["Item Name"])).rename(columns={"Item Name" : "Purchase Count"})

# The next column I want in my final table is "Average Purchase Price."
# To get the data, I need to first use groupby() to group "Price" by "Gender" and then find the average of the amounts by using mean().
# I then converted the groupby() data above into a data frame using pd.DataFrame.
# Since the data is now in a data frame format, I can append a new column labeled "Average Purchase Price" with the data found above.

purchase_analysis_by_gender_df["Average Purchase Price"] = pd.DataFrame(purchase_data.groupby(["Gender"]).mean()["Price"])["Price"]

# The next column I want in my final table is "Total Purchase Price."
# Similar to what I did above, to get the data, I need to first use groupby() to group "Price" by "Gender" and then find the total of the amounts by using sum().
# I then converted the groupby() data above into a data frame using pd.DataFrame.
# Since the data is now in a data frame format, I can append a new column labeled "Total Purchase Price" with the data found above.

purchase_analysis_by_gender_df["Total Purchase Value"] = pd.DataFrame(purchase_data.groupby(["Gender"]).sum()["Price"])

# The next column I want in my final table is "Avg Total Purchase per Person."
# Again, I will not use value_counts() because I want "Female" to be the first index.
# Since I am getting the average total per person, I need to use the data frame with unique player information. This is because a single player can have multiple purchases.
# I added another column to my purchase_analysis_by_gender_df data frame labeled "Gender Count" and added the values I found by using groupby() of "Price" by "Gender" in the unique_sn dataframe.

purchase_analysis_by_gender_df["Gender Count"] = pd.DataFrame(unique_sn.groupby(["Gender"]).count()["Price"])

# Now that I have the counts by gender added, I can divide the "Total Purchase Value" amounts by the total of unique players per gender.
# I created a new column labeled "Avg Total Purchase per Person."
# To get the average total purchase per person, I divided the values in the column "Total Purchase Value" by the values in the column "Gender Count."

purchase_analysis_by_gender_df["Avg Total Purchase per Person"] = purchase_analysis_by_gender_df["Total Purchase Value"]/purchase_analysis_by_gender_df["Gender Count"]

# Since I do not want the values in the "Gender Count" to show up in my final table, I used drop() to get rid of that column.

purchase_analysis_by_gender_df = purchase_analysis_by_gender_df.drop(columns=["Gender Count"])

# I used style.format() on my data frame so that the data looks clean.

purchase_analysis_by_gender_df = purchase_analysis_by_gender_df.style.format({"Average Purchase Price" : "${:.2f}", "Total Purchase Value" : "${:,.2f}", "Avg Total Purchase per Person": "${:.2f}"})

# Display the formatted table with the values.

purchase_analysis_by_gender_df

## Age Demographics: 


The following table will display the purchase count, average purchase price, total purchase value, and average purchase total per person by the following age groups: <10, 10-14, 15-19, 20-24, 25-29, 30-24, 35-39, and 40+.

In [6]:
# To get the bins in which the data will be held, I created a list of the cut off ages.

bins = [0, 9, 14, 19, 24, 29, 34, 39, 46]

# I also need to create the names for the bins. There will be one less name than the items in the bins.

group_names = ["<10", "10-14", "15-19", "20-24", "25-29", "30-34", "35-39", "40+"]

# I created a data frame of the unique players by dropping the duplicates using drop_duplicates(). This is the same as above. I just chose to create a new variable.

unique_sn_df = purchase_data.drop_duplicates(["SN"])

# To categorize the existing players using the age bins, I used pd.cut() to sort the "Age" values by the bin names.

unique_sn_df["Age Group"] = pd.cut(unique_sn["Age"], bins, labels=group_names)

# To get a data frame that shows the total count of players by age group, I first used groupby() and count() to group the "Item Name" by "Age Group." I can actually use any column, but I chose "Item Name."
# I then used pd.DataFrame to convert the above data into a data frame.
# I need to rename the "Item Name" to "Total Count" since I am finding the total count of players.

unique_sn_group_df = (pd.DataFrame(unique_sn_df["Item Name"].groupby(unique_sn_df["Age Group"]).count())).rename(columns = {"Item Name" : "Total Count"})

# Since I want to find the percentage of players by age group, I first used a list comprehension to take the value in the "Total Count" column above and divided each value by the total unique players found earlier (sum_players).
# While in the loop, I also decided to format the values by using .format().
# From the list comprehension, I now have a list of values that I can add to my data frame.
# I added a new column to the unique_sn_group_df data frame labeled "Percentage of Players" by setting it equal to the values I calculated.

unique_sn_group_df["Percentage of Players"] = ["{:.2f}".format((i/sum_players)*100) for i in unique_sn_group_df["Total Count"]]

# Display the formatted table with the values.

unique_sn_group_df    

C:\Users\Carolyn\Anaconda3\lib\site-packages\ipykernel_launcher.py:15: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  from ipykernel import kernelapp as app


Total Count Percentage of Players
Age Group                                   
<10                 17                  2.95
10-14               22                  3.82
15-19              107                 18.58
20-24              258                 44.79
25-29               77                 13.37
30-34               52                  9.03
35-39               31                  5.38
40+                 12                  2.08

## Purchasing Analysis (Age): 


The following table will display the purchase count, average purchase price, total purchase value, and average purchase per person by age group.

In [7]:
# Since I need to find the number of purchases made in each age group, I will be using the full data and not the unique table created earlier.
# Using the same bins and bin names created above, I used pd.cut() to categorize the values in the "Age" column into the designated age group.

purchase_data["Age Group"] = pd.cut(purchase_data["Age"], bins, labels=group_names)

# From the table created above, I used groupby() and count() to calculate the purchase count in each age group. 
# Since I am just getting a count, I can actually use any column. I chose the "Item Name" column here. 
# I will be adding values later to this table so I need the column title to be accurate. Using rename(), I renamed "Item Name" to "Purchase Count."

purchase_data_df = (pd.DataFrame(purchase_data["Item Name"].groupby(purchase_data["Age Group"]).count())).rename(columns = {"Item Name" : "Purchase Count"})

# The next column I want to add is the "Average Purchase Price."
# Similar to what I did previously, I took the "Price" values from the entire purchase data and grouped them by the age groups. 
# After the "Price" values are grouped by age group, I used mean() to find the average of the values per age group.
# I converted the data into a data frame using pd.DataFrame().
# Once the data frame is created, I can add a new column to my purchase_data_df with the column name "Average Purchase Price."

purchase_data_df["Average Purchase Price"] = (pd.DataFrame(purchase_data["Price"].groupby(purchase_data["Age Group"]).mean()))

# The next column I want to add is the "Total Purchase Price."
# To get the total purchase value, I used groupby() to again, group the "Price" values by the age groups.
# I used sum() to get a total of the "Price" values in each age group and set the data to a data frame using pd.DataFrame().
# From that data frame, I added the values to my purchase_data_df and titled the column "Total Purchase Value."

purchase_data_df["Total Purchase Value"] = pd.DataFrame(purchase_data["Price"].groupby(purchase_data["Age Group"]).sum())

# The next column is the "Avg Total Purchase per Person."
# To get the values for this column, I took the values listed in the "Total Purchase Value" column of the purchase_data_df and divided those values by the values in the "Total Count" column of the unique_sn_group_df found earlier.
# I used the "Total Count" values from the unique data frame because I am getting the average total purchase price per person. It wouldn't be accurate if I used the entire count as some players bought multiple items.
# I added the values calculated to a new column titled "Avg Total Purchase per Person" in purchase_data_df.

purchase_data_df["Avg Total Purchase per Person"] = purchase_data_df["Total Purchase Value"]/unique_sn_group_df["Total Count"]

# To get my table to display the purchase values as $0.00, I used style.format().

formatted_purchase_data_df = purchase_data_df.style.format({"Average Purchase Price" : "${:.2f}", "Total Purchase Value" : "${:,.2f}", "Avg Total Purchase per Person" : "${:.2f}"})

# Display the formatted table with the values.

formatted_purchase_data_df

## Top Spenders: 


The following table will display the top 5 spenders in the game by total purchase value. The table will also display the SN, purchase count, average purchase price, and total purchase value for each of the top spenders.

In [8]:
# To get the top spenders, I first need to count how many items each player bought. 
# To do this, I used groupby() and count() to count how many items each "SN" player bought. Here, I can use any column but I chose "Price."
# I set the data to a data frame and renamed the "Price" column to "Purchase Count" to get the table I will be adding values to.

top_spenders_df = (pd.DataFrame(purchase_data["Price"].groupby(purchase_data["SN"]).count())).rename(columns = {"Price" : "Purchase Count"})

# The next column I want to add to my top_spenders_df is the "Average Purchase Price."
# To get the average, I used groupby() and mean() to group the "Price" values by "SN" or players and finding the average of those values.
# I added the averages found to top_spenders_df under the "Average Purchase Price" column.

top_spenders_df["Average Purchase Price"] = (pd.DataFrame(purchase_data["Price"].groupby(purchase_data["SN"]).mean()))

# The next column I want to add is the "Total Purchase Value."
# To get a total for each user, I used groupby() and sum() to group the "Price" values by "SN" or players and added the values to get a total.
# I added the sums to a new column in top_spenders_df labeled "Total Purchase Value."

top_spenders_df["Total Purchase Value"] = (pd.DataFrame(purchase_data["Price"].groupby(purchase_data["SN"]).sum()))

# Since I want the table to display the top 5 spenders in the game by total purchase value, I used sort_values() to sort the values in "Total Purchase Value" in descending order (ascending needs to be set to False).

top_spenders_df = top_spenders_df.sort_values(["Total Purchase Value"], ascending = False)

# To get my table to display the purchase values as $0.00, I need to first convert the values in columns "Average Purchase Price" and "Total Purchase Value" to a float.
# I used astype() to convert the values to float.
# I then used map() and format() to format the values to display as "${:.2f}".

top_spenders_df["Average Purchase Price"] = top_spenders_df["Average Purchase Price"].astype(float).map("${:.2f}".format)
top_spenders_df["Total Purchase Value"] = top_spenders_df["Total Purchase Value"].astype(float).map("${:.2f}".format)

# To display the formatted values for the top 5 spenders, I used head().

top_spenders_df.head()

Purchase Count Average Purchase Price Total Purchase Value
SN                                                                     
Lisosia93                 5                  $3.79               $18.96
Idastidru52               4                  $3.86               $15.45
Chamjask73                3                  $4.61               $13.83
Iral74                    4                  $3.40               $13.62
Iskadarya95               3                  $4.37               $13.10

## Most Popular Items: 


The following table will display the 5 most popular items by purchase count. The table will also display the item ID, item name, purchase count, item price, and total purchase value for each of the most popular items.

In [9]:
# To get the most popular items, I need to use the entire purchase data and count how many times an item was purchased.
# I used groupby() to group the data by "Item ID," "Item Name," and "Price" and count() to get the number of times the item was purchased. I can use any column so I chose "Gender."
# I chose to group "Price" as well so that I can use the data collected there as a separate column for later.
# I then used sort_values() to sort the "Gender" counts with the highest count at the top (ascending = False) of the list in my table.
# Since "Gender" is not an appropriate title for the values collected, I used rename() to rename "Gender" to "Purchase Count."

most_popular_df = pd.DataFrame(purchase_data.groupby(["Item ID", "Item Name", "Price"]).count()["Gender"]).sort_values("Gender", ascending = False).rename(columns = {"Gender" : "Purchase Count"})

# With the data sorted, I used reset_index() to move the "Price" values out of the index so that it is it's own column.
# I used rename() to change the title from "Price" to "Item Price."

most_popular_df = (most_popular_df.reset_index(level="Price")).rename(columns = {"Price" : "Item Price"})

# To get the total purchase value, I multiplied the values under the "Item Price" column by the values under the "Purchase Count" column.
# I set those values to a new column titled "Total Purchase Value."

most_popular_df["Total Purchase Value"] = most_popular_df["Item Price"]*most_popular_df["Purchase Count"]

# To organize the data, I used reindex() to change the order of the columns.
# I also created two names for the same dataframe because I want to use the unformatted data frame later to sort the table by most profitable items.
# I can reset the format in order to sort, but setting the name ahead of time is much quicker.

most_popular_unformatted_df = most_popular_df.reindex(columns=["Purchase Count", "Item Price", "Total Purchase Value"])

most_popular_formatted_df = most_popular_df.reindex(columns=["Purchase Count", "Item Price", "Total Purchase Value"])

# To get my table to display the purchase values as $0.00, I need to first convert the values in columns "Item Price" and "Total Purchase Value" to a float.
# I used astype() to convert the values to float.
# I then used map() and format () to format the values to display as "${:.2f}".

most_popular_formatted_df["Item Price"] = most_popular_formatted_df["Item Price"].astype(float).map("${:.2f}".format)
most_popular_formatted_df["Total Purchase Value"] = most_popular_formatted_df["Total Purchase Value"].astype(float).map("${:.2f}".format)

# To display the formatted values for the top 5 spenders, I used head().

most_popular_formatted_df.head()

Purchase Count  \
Item ID Item Name                                                      
178     Oathbreaker, Last Hope of the Breaking Storm              12   
145     Fiery Glass Crusader                                       9   
108     Extraction, Quickblade Of Trembling Hands                  9   
82      Nirvana                                                    9   
19      Pursuit, Cudgel of Necromancy                              8   

                                                     Item Price  \
Item ID Item Name                                                 
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
145     Fiery Glass Crusader                              $4.58   
108     Extraction, Quickblade Of Trembling Hands         $3.53   
82      Nirvana                                           $4.90   
19      Pursuit, Cudgel of Necromancy                     $1.02   

                                                     Total Purchase Value  
Item ID Item Name                                                          
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
145     Fiery Glass Crusader                                       $41.22  
108     Extraction, Quickblade Of Trembling Hands                  $31.77  
82      Nirvana                                                    $44.10  
19      Pursuit, Cudgel of Necromancy                               $8.16

## Most Profitable Items: 


The following table will display the 5 most profitable items by total purchase value. The table will also display the item ID, item Name, purchase count, item price, and total purchase value for each of the most profitable items.

In [10]:
# To find the most profitable items, I need to sort the "Total Purchase Value" column in descending order.
# In the section above, I set an unformatted version of my data frame as most_popular_unformatted_df. 
# I used sort_values() to sort the values in the "Total Purchase Value" column in descending order (ascending = False).
# I could have formatted the values back to sortable data but decided to just set another name for the data frame above.

most_profitable_df = most_popular_unformatted_df.sort_values(["Total Purchase Value"], ascending = False)

# To get my table to display the purchase values as $0.00, I need to first convert the values in columns "Item Price" and "Total Purchase Value" to a float.
# I used astype() to convert the values to float.
# I then used map() and format () to format the values to display as "${:.2f}".

most_profitable_df["Item Price"] = most_profitable_df["Item Price"].astype(float).map("${:.2f}".format)
most_profitable_df["Total Purchase Value"] = most_profitable_df["Total Purchase Value"].astype(float).map("${:.2f}".format)

# To display the formatted values for the top 5 most profitable items, I used head().

most_profitable_df.head()

Purchase Count  \
Item ID Item Name                                                      
178     Oathbreaker, Last Hope of the Breaking Storm              12   
82      Nirvana                                                    9   
145     Fiery Glass Crusader                                       9   
92      Final Critic                                               8   
103     Singed Scalpel                                             8   

                                                     Item Price  \
Item ID Item Name                                                 
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
82      Nirvana                                           $4.90   
145     Fiery Glass Crusader                              $4.58   
92      Final Critic                                      $4.88   
103     Singed Scalpel                                    $4.35   

                                                     Total Purchase Value  
Item ID Item Name                                                          
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
82      Nirvana                                                    $44.10  
145     Fiery Glass Crusader                                       $41.22  
92      Final Critic                                               $39.04  
103     Singed Scalpel                                             $34.80

## Data Analysis Report

Based on the data collected above, there are a few observable trends:

1.	The first observable trend is that the majority of players in this game are male. Out of all the players, 84.03% are male, 14.06% are female, and 1.91% are listed as other/non-disclosed. The total amount of male players (484) is significantly more than the female and other/non-disclosed players combined (81+11 = 92).


2.	The majority of players fall in the 20-24 age group. Players that fall in the 20-24 age group make up 44.79% of total players. That is nearly half of the number of players in this game. After that age group, players that fall in the 15-19 age group make up 18.58% of players. 


3.	Item 178 or “Oathbreaker, Last Hope of the Breaking Storm” is the most popular item in the game as it is purchased the most often (12 times). It is also the most profitable item in the game with the total purchase value being 50.76 dollars. Though this item costs more than the average price of items in the game, it is still the most popular item among the players. The item costs 4.23 dollars while the average price of items in the game is 3.05 dollars.
